File: crates/transport/src/dto/entity_resolution.rs
use serde::{Deserialize, Serialize};

/// A resolved entity after canonicalization and merge.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ResolvedEntityDto {
    pub id: String,
    pub entity_type: String,
    pub canonical_value: String,
    pub source_entities: Vec<String>,
    pub source_count: u32,
    pub confidence: f64,
    pub attributes: Vec<String>,
}

/// Result of a deduplication merge pass.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EntityMergeResultDto {
    pub resolved_count: u32,
    pub merged_count: u64,
    pub resolved: Vec<ResolvedEntityDto>,
}

/// The type of an inferred entity-to-entity relationship.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum EntityRelationshipTypeDto {
    CommunicatesWith,
    Owns,
    LoggedInto,
    Executed,
    Downloaded,
    Accessed,
}

/// An inferred relationship between two entities, discovered from graph edge
/// patterns during entity resolution Phase 2.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EntityRelationshipDto {
    pub id: String,
    pub case_id: String,
    pub source_entity_id: String,
    pub target_entity_id: String,
    pub relationship_type: EntityRelationshipTypeDto,
    pub confidence: f64,
    pub evidence_edge_ids: Vec<String>,
    pub created_at: String,
}

/// Result payload for entity relationship inference.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct EntityRelationshipResultDto {
    pub relationship_count: u32,
    pub relationships: Vec<EntityRelationshipDto>,
}

#[cfg(test)]
#[path = "../../tests/unit/dto/entity_resolution.rs"]
mod tests;

File: frontend/src/types/entityResolution.ts
export interface ResolvedEntity {
  id: string;
  entityType: string;
  canonicalValue: string;
  sourceEntities: string[];
  sourceCount: number;
  confidence: number;
  attributes: string[];
}

export interface EntityMergeResult {
  resolvedCount: number;
  mergedCount: number;
  resolved: ResolvedEntity[];
}

export type EntityRelationshipType =
  | 'communicatesWith'
  | 'owns'
  | 'loggedInto'
  | 'executed'
  | 'downloaded'
  | 'accessed';

export interface EntityRelationship {
  id: string;
  caseId: string;
  sourceEntityId: string;
  targetEntityId: string;
  relationshipType: EntityRelationshipType;
  confidence: number;
  evidenceEdgeIds: string[];
  createdAt: string;
}

export interface EntityRelationshipResult {
  relationshipCount: number;
  relationships: EntityRelationship[];
}
